File: Homework/Hang_Nguyen/lesson_6/house-robber-iii.py
# https://leetcode.com/problems/house-robber-iii/
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def rob(self, root):
        def robPlan(node):
            if not node: return (0, 0)
            left, right = robPlan(node.left), robPlan(node.right)
            now = node.val + left[1] + right[1]
            later = max(left) + max(right)
            return (now, later)
            
        return max(robPlan(root))


# TEST CASE
solution = Solution()

left = TreeNode(2, None, TreeNode(3))
right = TreeNode(3, None, TreeNode(1))
root = TreeNode(3, left, right)

print("TEST 1 is " + str(solution.rob(root)))

left = TreeNode(4, TreeNode(1), TreeNode(3))
right = TreeNode(5, None, TreeNode(1))
root = TreeNode(3, left, right)

print("TEST 2 is " + str(solution.rob(root)))

File: Homework/Duy_Nguyen/day_7/bfs/HouseRobberIII.java
package org.global.dev.day_7.bfs;

import org.global.dev.common.TreeNode;

public class HouseRobberIII {
	
	/**
	 * https://leetcode.com/problems/house-robber-iii/?fbclid=IwAR0hSqeGKZJr4cunLzKAkVnSzkdxR9it6A7noc1VNHRo3nTU8rzXIhKxmQQ
	 */
	
	public int rob(TreeNode root) {
		return helperRecursive(root, false);
	}
	
	public int helperRecursive(TreeNode root, boolean isParentRobbed) {
		if (root == null) {
			return 0;
		}
		if (isParentRobbed) {
			// Can not rob this;
			return helperRecursive(root.left, false) + helperRecursive(root.right, false);
		} else {
			// There two choice rob or not rob this
			int chooseRob = root.val + helperRecursive(root.right, true) + helperRecursive(root.left, true);
			int notRob = helperRecursive(root.right, false) + helperRecursive(root.left, false);
			return Math.max(chooseRob, notRob);
		}
	}
	
	public int robOp(TreeNode root) {
		int[] res = helperOptimize(root);
		return Math.max(res[0], res[1]);
	}
	
	public int[] helperOptimize(TreeNode root) {
		// return [rob this node, not rob this node]
		if (root == null) {
			return new int[] { 0, 0 };
		}
		
		int[] left = helperOptimize(root.left);
		int[] right = helperOptimize(root.right);
		
		int notRob = Math.max(left[1], left[0]) + Math.max(right[0], right[1]);
		int rob = root.val + left[1] + right[1];
		
		return new int[]{rob, notRob};
	}
	
	
}
